# Exploring Hacker News Posts

---



Hacker News is a site where user-submitted stories (known as "posts") receive votes and comments, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of the Hacker News listings can get hundreds of thousands of visitors as a result

We're specifically interested in posts with titles that begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just something interesting

We'll compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

In [ ]:
opened_file = open('hacker_news.csv')
from csv import reader
read_file = reader(opened_file)
hn = list(read_file)

extract header:

In [ ]:
headers=hn[0]
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [ ]:
hn=hn[1:]
hn[0:3]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20']]

Creating three lists called ask_posts, show_posts, and other_posts:

In [ ]:
ask_posts=[]
show_posts=[]
other_posts=[]

In [ ]:
for post in hn:
  title=post[1]
  if title.lower().startswith("ask hn"):
    ask_posts.append(post)
  elif title.lower().startswith("show hn"):
    show_posts.append(post)
  else:
    other_posts.append(post)

In [ ]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [ ]:
total_ask_comments=0
for row in ask_posts:
  num_comments=row[4]
  num_comments=int(num_comments)
  total_ask_comments+=num_comments
avg_ask_comments= total_ask_comments/ len(ask_posts)

total_show_comments=0
for row in show_posts:
  num_comments=row[4]
  num_comments=int(num_comments)
  total_show_comments+=num_comments
avg_show_comments= total_show_comments/ len(show_posts)

print(avg_ask_comments)
print(avg_show_comments)
print(avg_ask_comments-avg_show_comments)


14.038417431192661
10.31669535283993
3.7217220783527303


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10.
Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

* Calculate the number of ask posts created in each hour of the day, along with the number of comments received.
* Calculate the average number of comments ask posts receive by hour created.

In [ ]:
import datetime as dt

In [ ]:
result_list=[]
for post in ask_posts:
  result_list.append([post[6], int(post[4])])

counts_by_hour= { }
comments_by_hour= { }
date_format = "%m/%d/%Y %H:%M"

for post in result_list[1:]:
  date = post[0]
  comment = post[1]
  time = dt.datetime.strptime(date, date_format).strftime("%H")
  if time in counts_by_hour:
      comments_by_hour[time] += comment
      counts_by_hour[time] += 1
  else:
      comments_by_hour[time] = comment
      counts_by_hour[time] = 1


In [ ]:
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr,comments_by_hour[hr]/counts_by_hour[hr]])

avg_by_hour

[['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['09', 5.568181818181818],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

In [ ]:
swap_avg_by_hour=[ ]
for row in avg_by_hour:
  swap_avg_by_hour.append([row[1],row[0]])
print(swap_avg_by_hour)

[[14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [5.568181818181818, '09'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [ ]:
sorted_swap=sorted(swap_avg_by_hour, reverse=True)

In [ ]:
print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
  print("{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg) )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post



The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post.

# Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).